platform/lang-api/src/com/intellij/execution/actions/PreferredProducerFind.java

```diff
@@ -140,4 +157,38 @@ private static ConfigurationFromContext findConfigurationFromContext(final Locat
     }
     return null;
   }
+
+  @Nullable
+  private static List<ConfigurationFromContext> getConfigurationsFromAlternativeLocations(
+    @NotNull MultipleRunLocationsProvider.AlternativeLocationsInfo alternativeLocationsInfo,
+    @NotNull Location originalLocation,
+    boolean strict
+  ) {
+    SmartList<ConfigurationFromContext> result = new SmartList<>();
+    for (Location alternativeLocation : alternativeLocationsInfo.getAlternativeLocations()) {
+      ConfigurationContext fakeContextForAlternativeLocation = new ConfigurationContext(alternativeLocation);
+      List<ConfigurationFromContext> configurationsForLocation =
+        doGetConfigurationsFromContext(alternativeLocation, fakeContextForAlternativeLocation, strict);
+      if (configurationsForLocation != null) {
+        for (ConfigurationFromContext configurationFromContext : configurationsForLocation) {
+          configurationFromContext.setFromAlternativeLocation(true);
+          String locationDisplayName = alternativeLocationsInfo.getProvider().getLocationDisplayName(alternativeLocation, originalLocation);
+          configurationFromContext.setAlternativeLocationDisplayName(
+            locationDisplayName
+          );
+          RunConfiguration configuration = configurationFromContext.getConfiguration();
+          if (configuration instanceof LocatableConfigurationBase && ((LocatableConfiguration)configuration).isGeneratedName()) {
+            configuration.setName(configuration.getName() + " " + locationDisplayName);
+            ((LocatableConfigurationBase)configuration).setNameChangedByUser(true);
```

Comment:
We need to handle renames of classes/methods from which these configurations have been created